### DISCUSSION TITLE
Open Modal from Server side.

### DISCUSSION THREAD

**Discussion by perkonto** [unpoly/unpoly#344] [Q&A]
Hello all!
I've started playing with Unpoly and I'm in the middle of the learning process.

How can I open a Modal/Drawer from the server?

For example, using the Unpoly Demo... If I navigate to the [Projects' list](https://demo.unpoly.com/projects) and I click in a Project I can see that the Modal opens and the url changes. But if I refresh the page (F5) or I navigate via [direct link](https://demo.unpoly.com/projects/766287), the modal is gone. 

I understand this behavior but I would like to open the Modal even when the page is refreshed.

Is it possible? is the **Server Protocol** the right direction to achieve this?

This is what I've tried so far without success:

```python
response.headers.set('X-Up-Target', "#myModal")
response.headers.set('X-Up-Mode', "modal")
```

Thanks

**Comment by adam12** [comment]
I don't believe it's possible. Setting `X-Up-Target` through a header is fine, but `X-Up-Mode` is a request header only.

**Comment by triskweline** [comment]
This is not possible right now.

It would not really help in your case. The browser is doing a full reload, so you need to render everything that is on the screen. In your case, the background layer *and* the modal over it.

If you prefer your modals to not show their URL in the address bar you can configure `up.layer.config.modal.history = false`. I usually prefer modals to open as a full page when reloaded.

**Reply by adam12** [reply]
> I usually prefer modals to open as a full page when reloaded.

I do as well.

**Comment by perkonto** [comment]
Thanks @adam12, @triskweline  for you quick answer.

And what about with `X-Up-Events`? Is it possible to achieve it with it?
Because I managed to trigger a custom event after user interaction but not after press F5 for example.

This is what I did so far:

```python
# Server:
response.headers.set('X-Up-Events', '[{ "type": "server:event", "hello": "world!" }]')
```

```js
// Client
up.on('server:event', function(event) {
    console.log(event.hello) // it prints world!
   // Open the modal with JS
});
```

This `server:event` is only triggered after Click `<a href="..." up-target="#myModal" up-layer="new">` but not when a Refresh the page (F5) or when I navigate via direct link.

**Reply by adam12** [reply]
You could probably render the page with your content inside a `<template>` tag or similar, and then call one of the `up.` functions on page load to create the modal.

I've done that in the past, but the modal being present wasn't critical for the page to be usable (it was a Welcome modal), so I am not sure I'd recommend the pattern for anything critical.

**Reply by perkonto** [reply]
Thanks @adam12 
Nothing critical. Just learning and trying to figure out what can & can't do with Unpoly.
